torchrec/csrc/dynamic_embedding/details/id_transformer_variant.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+#include <torchrec/csrc/dynamic_embedding/details/mixed_lfu_lru_strategy.h>
+#include <torchrec/csrc/dynamic_embedding/details/naive_id_transformer.h>
+#include <span>
+#include <type_traits>
+#include <variant>
+
+namespace torchrec {
+
+class LXUStrategyVariant {
+  // Currently, only MixedLFULRUStrategy is the LXUStrategyVariant. Add more
+  // strategy when it is necessary.
+  using Variant = std::variant<MixedLFULRUStrategy>;
+
+  // use to indicate VisitUpdator 's result
+  using _UpdateFunctor = std::function<MixedLFULRUStrategy::lxu_record_t(
+      std::optional<MixedLFULRUStrategy::lxu_record_t>,
+      int64_t,
+      int64_t)>;
+
+ public:
+  explicit LXUStrategyVariant(
+      const std::string& lxu_strategy_type,
+      uint16_t min_used_freq_power);
+
+  void update_time(uint32_t time);
+  template <typename T>
+  int64_t time(T record);
+
+  template <typename Visitor>
+  auto visit_updator(Visitor visit)
+      -> std::invoke_result_t<Visitor, _UpdateFunctor>;
+
+  template <typename Iterator>
+  std::vector<int64_t> evict(Iterator iterator, uint64_t num_to_evict);
+
+ private:
+  Variant strategy_;
+};
+
+class IDTransformerVariant {
+  // Currently, only NaiveIDTransformer is the IDTransformerVariant. Add more
+  // strategy when it is necessary.
+  using Variant = std::variant<NaiveIDTransformer<uint32_t>>;
+
```

Comment:
I think it would be simpler not to use std::variant, but just to rely on virtual functions instead. This way you would just have IDTransformer and NaiveIDTransformer, which extends IDTransformer. It also would have been much easier to reason about it.